### PR TITLE
fix race conditions when updating PlanningScene

### PR DESCRIPTION
**Pull request by rhaschke**
After having merged #63 we can finally attempt to merge https://github.com/rhaschke/moveit/pull/1 against ros-planning/moveit. This PR attempts to fix https://github.com/ros-planning/moveit_ros/issues/442 and the flaky test mentioned in https://github.com/ros-planning/moveit/pull/221#issuecomment-248223570. This PR draws on reverted PRs https://github.com/ros-planning/moveit_ros/pull/716 https://github.com/ros-planning/moveit_ros/pull/724, and https://github.com/ros-planning/moveit_ros/pull/728.

* I removed `AsyncSpinner` and `CallBackQueue` in `PlanningSceneMonitor` for now. With https://github.com/ros-planning/moveit_ros/pull/717 (#59) this is not required anymore to update the robot state. However `syncSceneUpdates()` relies on them, which is why I had to remove them for now too. It's anyway in discussion, whether this function is useful. I will file a separate PR for this.
* I renamed `CurrentStateMonitor::waitForCurrentState()` and added a deprecated fallback function calling the new method `waitForCompleteState()`. Deprecation should be OK in Kinetic, but not Indigo and Jade. This should be considered when back-porting.

### CODE REVIEW

File: moveit_ros/move_group/src/default_capabilities/execute_trajectory_service_capability.cpp [hunk 1]
```diff
@@ -86,14 +86,12 @@ bool move_group::MoveGroupExecuteService::executeTrajectoryService(moveit_msgs::
...
+        res.error_code.val = moveit_msgs::MoveItErrorCodes::TIMED_OUT;
       else
-        if (es == moveit_controller_manager::ExecutionStatus::PREEMPTED)
-          res.error_code.val = moveit_msgs::MoveItErrorCodes::PREEMPTED;
-        else
-          if (es == moveit_controller_manager::ExecutionStatus::TIMED_OUT)
-            res.error_code.val = moveit_msgs::MoveItErrorCodes::TIMED_OUT;
-          else
-            res.error_code.val = moveit_msgs::MoveItErrorCodes::CONTROL_FAILED;
+        res.error_code.val = moveit_msgs::MoveItErrorCodes::CONTROL_FAILED;
```
Comment (rhaschke) [inline moveit_ros/move_group/src/default_capabilities/execute_trajectory_service_capability.cpp]: This is simply reformatting nested if's into a linear structure.

File: moveit_ros/manipulation/move_group_pick_place_capability/src/pick_place_action_capability.cpp [hunk 2]
```diff
@@ -322,6 +322,7 @@ void move_group::MoveGroupPickPlaceAction::executePickupCallback(const moveit_ms
 {
   setPickupState(PLANNING);
 
+  context_->planning_scene_monitor_->waitForCurrentRobotState(ros::Time::now());
```
Comment (rhaschke) [inline moveit_ros/manipulation/move_group_pick_place_capability/src/pick_place_action_capability.cpp]: Before we start planning, ensure that we have the latest robot state received...

File: moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/current_state_monitor.h [hunk 3]
```diff
@@ -199,6 +208,7 @@ class CurrentStateMonitor
   ros::Time                                    last_tf_update_;
 
   mutable boost::mutex                         state_update_lock_;
+  mutable boost::condition_variable            state_update_condition_;
```
Comment (rhaschke) [inline moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/current_state_monitor.h]: This introduces an ABI incompatibility.

File: moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp [hunk 4]
```diff
@@ -878,7 +878,7 @@ bool TrajectoryExecutionManager::validate(const TrajectoryExecutionContext &cont
   ROS_DEBUG_NAMED("traj_execution", "Validating trajectory with allowed_start_tolerance %g", allowed_start_tolerance_);
 
   robot_state::RobotStatePtr current_state;
-  if (!csm_->waitForCurrentState(1.0) || !(current_state = csm_->getCurrentState()))
+  if (!csm_->waitForCurrentState(ros::Time::now()) || !(current_state = csm_->getCurrentState()))
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp]: Here, calling the new function waitForCurrentState() instead of the old one, which was renamed to waitForCompleteRobotState().

File: moveit_ros/move_group/src/default_capabilities/execute_trajectory_service_capability.cpp (hunk 1 above)
Comment (davetcoleman) [inline moveit_ros/move_group/src/default_capabilities/execute_trajectory_service_capability.cpp]: +1

File: moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/current_state_monitor.h [hunk 5]
```diff
@@ -136,11 +138,18 @@ class CurrentStateMonitor
...
   std::map<std::string, double> getCurrentStateValues() const;
 
-  /** @brief Wait for at most \e wait_time seconds until the complete current state is known. Return true if the full state is known */
-  bool waitForCurrentState(double wait_time) const;
+  /** @brief Wait for at most \e wait_time seconds for a robot state more recent than t */
+  bool waitForCurrentState(const ros::Time t=ros::Time::now(), double wait_time=1) const;
+
+  /** @brief Wait for at most \e wait_time seconds until the complete robot state is known. Return true if the full state is known */
+  bool waitForCompleteState(double wait_time) const;
+  /// replaced by waitForCompleteState, will be removed in L-turtle
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/current_state_monitor.h]: comment should be ``/** */`` for consistency

File: moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp [hunk 6]
```diff
@@ -818,6 +825,55 @@ void planning_scene_monitor::PlanningSceneMonitor::currentWorldObjectUpdateCallb
...
+  // However, scene updates are only published if the robot actually moves. Hence we need a timeout!
+  // As publishing planning scene updates is throttled (2Hz by default), a 1s timeout is a suitable default.
+  boost::shared_lock<boost::shared_mutex> lock(scene_update_mutex_);
+  ros::Time prev_robot_motion_time_ = last_robot_motion_time_;
+  while (last_robot_motion_time_ < t && // Wait until the state update actually reaches the scene.
+         timeout > ros::WallDuration())
+  {
+    ROS_DEBUG_STREAM_NAMED("planning_scene_monitor", "last robot motion: " << (t-last_robot_motion_time_).toSec() << " ago");
+    new_scene_update_condition_.wait_for(lock, boost::chrono::nanoseconds(timeout.toNSec()));
+    timeout -= ros::WallTime::now()-start; // compute remaining wait_time
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: spacing

**Comment by davetcoleman** [comment]
ping @v4hn  - this PR is the last blocker for the kinetic release!

also, @rhaschke do you mind rebasing for the merge conflicts?

**Comment by davetcoleman** [comment]
@rhaschke can you rebase this to help get this merged in faster?

File: moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp [hunk 7]
```diff
@@ -263,7 +263,24 @@ bool planning_scene_monitor::CurrentStateMonitor::haveCompleteState(const ros::D
...
+bool planning_scene_monitor::CurrentStateMonitor::waitForCurrentState(const ros::Time t, double wait_time) const
+{
+  ros::WallTime start = ros::WallTime::now();
+  ros::WallDuration elapsed;
+  ros::WallDuration timeout(wait_time);
+
+  boost::mutex::scoped_lock lock(state_update_lock_);
+  while (current_state_time_ < t)
+  {
+    state_update_condition_.wait_for(lock, boost::chrono::nanoseconds((timeout-elapsed).toNSec()));
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp]: space between ``-``

File: moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp [hunk 8]
```diff
@@ -263,7 +263,24 @@ bool planning_scene_monitor::CurrentStateMonitor::haveCompleteState(const ros::D
   return result;
 }
 
-bool planning_scene_monitor::CurrentStateMonitor::waitForCurrentState(double wait_time) const
+bool planning_scene_monitor::CurrentStateMonitor::waitForCurrentState(const ros::Time t, double wait_time) const
+{
+  ros::WallTime start = ros::WallTime::now();
+  ros::WallDuration elapsed;
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp]: explicitly initializing this to ``(0)`` rather than relying on the default constructor might make the code slightly easier to read

Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp]: explicitly initializing this to ``(0)`` rather than relying on the default constructor might make the code slightly easier to read

File: moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/current_state_monitor.h [hunk 9]
```diff
@@ -136,11 +138,18 @@ class CurrentStateMonitor
    *  @return Returns the map from joint names to joint state values*/
   std::map<std::string, double> getCurrentStateValues() const;
 
-  /** @brief Wait for at most \e wait_time seconds until the complete current state is known. Return true if the full state is known */
-  bool waitForCurrentState(double wait_time) const;
+  /** @brief Wait for at most \e wait_time seconds for a robot state more recent than t */
+  bool waitForCurrentState(const ros::Time t=ros::Time::now(), double wait_time=1) const;
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/current_state_monitor.h]: in comment explain what the return value means - e.g. ``@return false indicates the current robot state at time \e t was not found within the \e wait_time``

File: moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp [hunk 10]
```diff
@@ -263,7 +263,24 @@ bool planning_scene_monitor::CurrentStateMonitor::haveCompleteState(const ros::D
...
+  {
+    state_update_condition_.wait_for(lock, boost::chrono::nanoseconds((timeout-elapsed).toNSec()));
+    elapsed = ros::WallTime::now() - start;
+    if (elapsed > timeout)
+      return false;
+  }
+  return true;
+}
+
+bool planning_scene_monitor::CurrentStateMonitor::waitForCompleteState(double wait_time) const
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp]: So in this PR you took an old function ``waitForCurrentState`` and renamed it ``waitForCompleteState``? Then created a new function ``waitForCurrentState`` that collides with the old function's name? This sounds incredibly convoluted - shouldn't you retain the old function's name here and create a new function with new functionality under a new name??

File: moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp [hunk 11]
```diff
@@ -408,4 +433,7 @@ void planning_scene_monitor::CurrentStateMonitor::jointStateCallback(const senso
   if (update)
     for (std::size_t i = 0 ; i < update_callbacks_.size() ; ++i)
       update_callbacks_[i](joint_state);
+
+  // notify *after* potential update callbacks
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp]: specify "notify waitForCurrentState() *after* potential..."

File: moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp [hunk 12]
```diff
@@ -1139,8 +1198,9 @@ void planning_scene_monitor::PlanningSceneMonitor::updateSceneWithCurrentState()
 
     {
       boost::unique_lock<boost::shared_mutex> ulock(scene_update_mutex_);
+      last_update_time_ = last_robot_motion_time_ = current_state_monitor_->getCurrentStateTime();
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: I'm not a fan of multiple variable assignments on one line, personally. Its more difficult to parse for the reader.

File: moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp [hunk 13]
```diff
@@ -818,6 +825,55 @@ void planning_scene_monitor::PlanningSceneMonitor::currentWorldObjectUpdateCallb
...
+  ros::WallDuration timeout(wait_time);
+
+  ROS_DEBUG_NAMED("planning_scene_monitor", "sync robot state to: %.3fs", fmod(t.toSec(), 10.));
+
+  if (current_state_monitor_)
+  {
+    // Wait for next robot update in state monitor. Those updates are only passed to PSM when robot actually moved!
+    enforce_next_state_update_ = true; // enforce potential updates to be directly applied
+    bool success = current_state_monitor_->waitForCurrentState(t, wait_time);
+    enforce_next_state_update_ = false; // enforce potential updates to be directly applied
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: comment: "enforce potential updates to *not* be directly applied"

File: moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp [hunk 14]
```diff
@@ -818,6 +825,55 @@ void planning_scene_monitor::PlanningSceneMonitor::currentWorldObjectUpdateCallb
...
+
+    /* If the robot doesn't move, we will never receive an update from CSM in planning scene.
+       As we ensured that an update, if it is triggered by CSM, is directly passed to the scene,
+       we can early return true here (if we successfully received a CSM update). Otherwise return false. */
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: nice comment!

File: moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp [hunk 15]
```diff
@@ -818,6 +825,55 @@ void planning_scene_monitor::PlanningSceneMonitor::currentWorldObjectUpdateCallb
...
+
+    ROS_WARN("Failed to fetch current robot state.");
+    return false;
+  }
+
+  // Sometimes there is no state monitor. In this case state updates are received as part of scene updates only.
+  // However, scene updates are only published if the robot actually moves. Hence we need a timeout!
+  // As publishing planning scene updates is throttled (2Hz by default), a 1s timeout is a suitable default.
+  boost::shared_lock<boost::shared_mutex> lock(scene_update_mutex_);
+  ros::Time prev_robot_motion_time_ = last_robot_motion_time_;
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: there should not be an underscore at the end of this local variable

**Comment by rhaschke** [comment]
Rebased.

File: moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp (hunk 10 above)
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp]: thinking about this more  - i really don't think this is a good idea. you are changing core functionality of a function in a public API that others might be using in their own projects (albeit maybe unlikely) in a way that is very non-obvious

File: moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h [hunk 16]
```diff
@@ -513,7 +522,7 @@ class PlanningSceneMonitor : private boost::noncopyable
 
   /// Last time the state was updated from current_state_monitor_
   // Only access this from callback functions (and constructor)
-  ros::WallTime last_state_update_;
+  ros::WallTime wall_last_state_update_;
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h]: also this should be in a separate PR to keep this PR easier to review

File: moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp [hunk 17]
```diff
@@ -827,9 +883,9 @@ void planning_scene_monitor::PlanningSceneMonitor::lockSceneRead()
 
 void planning_scene_monitor::PlanningSceneMonitor::unlockSceneRead()
 {
-  scene_update_mutex_.unlock_shared();
   if (octomap_monitor_)
     octomap_monitor_->getOcTreePtr()->unlockRead();
+  scene_update_mutex_.unlock_shared();
```
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: changes like this should be in a separate PR

File: moveit_ros/planning_interface/test/python_move_group.py [hunk 18]
```diff
@@ -56,6 +56,15 @@ def test_validation(self):
...
 
+    def test_robot_state_update(self):
+        current = np.asarray(self.group.get_current_joint_values())
+        for i in range(3):
+            target = current + np.random.uniform(-0.5, 0.5, size = current.shape)
+            # if plan was successfully executed, current state should be reported at target
+            if self.group.execute(self.plan(target)):
+                 actual = np.asarray(self.group.get_current_joint_values())
+                 self.assertTrue(np.allclose(target, actual, atol=1e-4, rtol=0.0))
+
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/python_move_group.py]: i think this could be in a separate PR also

File: moveit_ros/move_group/src/default_capabilities/execute_trajectory_service_capability.cpp (hunk 1 above)
Comment (davetcoleman) [inline moveit_ros/move_group/src/default_capabilities/execute_trajectory_service_capability.cpp]: changes like this should be in a separate PR

**Comment by davetcoleman** [comment]
I really want to get this merged in so we can release kinetic, so have attempted to review, above. Would greatly appreciate anyone who can test this PR on their physical robot setup. @Jntzko?

**Comment by v4hn** [comment]
Thanks dave!

> waitForCurrentState

Yes this is meddling with a core function, but I discussed that with @rhaschke in the old request already.
The old function is plainly broken beyond repair! While it is called `waitForCurrentState` it does not wait for a current state at all! @rhaschke proposed to change this (for the better) and provide an old fallback and I fully agree with him here.
A number of people rely on that function to do what it says, and it should do that.

**Comment by Jntzko** [comment]
I've tested this PR in indigo with the 9 commits from @rhaschke in the simulation.

Therefore I made a small test case where the ur5 moves in a loop to four positions.
After a few iterations I get this error:

`Invalid Trajectory: start point deviates from current robot state more than 0.01
joint 'ur5_wrist_3_joint': expected: -2.52336, current: -2.53886`

As @v4hn told me, this error should not appear with the fixes from the PR. 

To get this test running you need to check out the following repositories:

  The branch testFix#442 from:
  https://github.com/Jntzko/moveit.git

  The test case:
  https://github.com/Jntzko/fix_haschke.git

  And also this repositories for our lab setup: 
  https://github.com/TAMS-Group/tams_ur5.git
  https://github.com/TAMS-Group/tams_ur5_setup.git

Start the test with:
`roslaunch tams_ur5_setup_moveit_config demo.launch`
and
`rosrun fix_haschke move_loop`

**Comment by rhaschke** [comment]
I guess, to reproduce the test, I also would need to have a UR arm, wouldn't I?
Is it possible that those deviations originate from sensor noise (or breaks)? 
Actually, I added a similar test in [moveit_ros/planning_interface/test/python_move_group.py](https://github.com/ros-planning/moveit/pull/232/files#diff-700c130c2b0b0f46b7115e123b9c8d61) doing some (only 3) random motions in simulations. @Jntzko, could you please run this test too, possibly increasing the number of loops? It runs like this:
rostest moveit_ros_planning_interface python_move_group.test

This test runs without problems for me with a simulation-only robot.

**Comment by v4hn** [comment]
On Mon, Oct 10, 2016 at 06:02:32AM -0700, Robert Haschke wrote:
> I guess, to reproduce the test, I also would need to have a UR arm, wouldn't I?

No, he ran this in demo mode. That's why I'm pretty sure the problem lies in the synchronization.

**Comment by Jntzko** [comment]
@rhaschke I've tested your code, also just in the simulation and I get the same "start point deviates from current robot state" error.

**Comment by davetcoleman** [comment]
This discussion on what tests you are using to check this PR inspired me to write a stub tutorial on the topic: https://github.com/ros-planning/moveit_tutorials/pull/24

File: moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h [hunk 19]
```diff
@@ -344,6 +344,13 @@ class PlanningSceneMonitor : private boost::noncopyable
   /** @brief This function is called every time there is a change to the planning scene */
   void triggerSceneUpdateEvent(SceneUpdateType update_type);
 
+  /** \brief Wait for robot state to become more recent than t.
+   *
+   * If there is no state monitor active, there will be no scene updates.
+   * Hence, you can specify a timeout to wait for those updates. Default is 1s.
+   */
+  bool waitForCurrentRobotState(const ros::Time &t, double wait_time = 1.);
```
Comment (rhaschke) [inline moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h]: As explained by @v4hn, we agreed to correctly rename the old function waitForCurrentRobotState -> waitForCompleteRobotState to reflect its actual semantics.
The new function waitForCurrentRobotState actually does what its name states: waiting for an update of the current robot state.

File: moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h [hunk 20]
```diff
@@ -416,6 +423,9 @@ class PlanningSceneMonitor : private boost::noncopyable
   planning_scene::PlanningSceneConstPtr scene_const_;
   planning_scene::PlanningScenePtr      parent_scene_; /// if diffs are monitored, this is the pointer to the parent scene
   boost::shared_mutex                   scene_update_mutex_; /// mutex for stored scene
+  ros::Time                             last_update_time_; /// Last time the state was updated
```
Comment (rhaschke) [inline moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h]: I didn't want to change the name. I just moved all related variables to one location.
As PlanningSceneMonitor deals with scene updates only, the naming is unique already, isn't it?

File: moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp (hunk 10 above)
Comment (rhaschke) [inline moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp]: As explained by @v4hn, this cleans up the API, naming function according to their semantics...

File: moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp (hunk 15 above)
Comment (rhaschke) [inline moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: Only member variables should have an underscore in the end, right?

File: moveit_ros/planning_interface/test/python_move_group.py (hunk 18 above)
Comment (rhaschke) [inline moveit_ros/planning_interface/test/python_move_group.py]: Why this? This is the unittest which tests the behavior fixed by this PR.

**Comment by rhaschke** [comment]
> I've tested your code, also just in the simulation and I get the same "start point deviates from current robot state" error.

@Jntzko Sorry, but I cannot reproduce the error in Kinetic with my proposed unit test. Could you, please, send me the whole output of running the unit test?

`rostest moveit_ros_planning_interface python_move_group.test`

**Comment by rhaschke** [comment]
I addressed the additional comments from @davetcoleman.

File: moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h (hunk 20 above)
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h]: That's exactly what is confusing - the naming is not unique because the comment indicates it deals with *state* updates, not scene updates as one might assume from the class name PlanningSceneMonitor

File: moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp (hunk 15 above)
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: right, and ``prev_robot_motion_time_`` is a local variable

File: moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp (hunk 10 above)
Comment (davetcoleman) [inline moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp]: ok

File: moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp (hunk 15 above)
Comment (rhaschke) [inline moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: Oh. I missed the *not* in your original comment.

**Comment by v4hn** [comment]
We will try to look at this tomorrow.
There is no sense in merging this if it actually fails to achieve what it was written for.
And it does not make too much sense to me that it works fine for @rhaschke on kinetic but fails for @Jntzko on indigo.

**Comment by rhaschke** [comment]
> There is no sense in merging this if it actually fails to achieve what it was written for.

Agreed. However, I don't see a reason yet, why it should fail. Do you?

> And it does not make too much sense to me that it works fine for @rhaschke on kinetic but fails for @Jntzko on indigo.

That's why I asked for the full log.

**Comment by rhaschke** [comment]
@Jntzko Please send the full log of the unit test to allow me to further look into this.

**Comment by Jntzko** [comment]
@rhaschke I have cherry picked your latest commits in my indigo branch too and ran the test multiple times.

The following error does not appear in every run within the 30 movements, but it still appears every now and then(appeared in 2 of 3 runs):
> [ERROR] [1476803281.592726674]: 
Invalid Trajectory: start point deviates from current robot state more than 0.01
joint 'joint_6': expected: -0.359403, current: -0.37355

I work with the `testFix#442` branch of this repo:
https://github.com/Jntzko/moveit.git

Here's one of the rostest-log files where the error appeared:
[rostest-tams121-26002.log.txt](https://github.com/ros-planning/moveit/files/536883/rostest-tams121-26002.log.txt)

**Comment by Jntzko** [comment]
@rhaschke I ran your test on kinetic now and increased the number of movements to 100. 
When I ran the test the third time, the error appeared again:
>[ERROR] [1476828218.417227953]: 
Invalid Trajectory: start point deviates from current robot state more than 0.01
joint 'joint_1': expected: -0.345288, current: -0.145362

Here's the rostest-log file:
[rostest-thinkpad-28045.log.txt](https://github.com/ros-planning/moveit/files/537483/rostest-thinkpad-28045.log.txt)

**Comment by rhaschke** [comment]
I could indeed reproduce the problem in Kinetic too now. The update is missed rarely but sometimes. I think, the core issue is that the `TrajectoryExecutionManager` signals finishing of trajectory execution as soon as the last controller values were send out. However, the controller will need some time to process them and _only after some delay_ the final joint state will be published by the hardware or controller.

We have the same situation with fake controllers too: They publish to `/move_group/fake_controller_joint_states` and a `joint_state_publisher` re-publishes with a fixed rate and its _own timestamp_.

Hence, it might happen, that `waitForCurrentRobotState(ros::Time::now())` is called with a threshold time earlier than the finally published `joint_states`. In that case we will miss these final updates. Modifying joint_state_publisher to re-publish with the received timestamp (instead of a new one), solved the problem on my side, which supports the explanation.

However, this is of course not a feasible solution. The core issue is that we cannot know the delay between the last controller command send by `TrajectoryExecutionManager` and final `joint_states` received. A potential solution could be to wait in `TrajectoryExecutionManager` until the robot "settles" before sending the finished signal. As I integrated a `CurrentStateMonitor` recently, that should be feasible - using the same noise threshold as for `validate()`.

Hence, I suggest to drop this PR for Kinetic. However, I will file some parts of this PR as separate PRs.
